### PR TITLE
[Collectd 6] gpu_sysman: add "LogMetrics" option

### PR DESCRIPTION
**Pull request by eero-t**
ChangeLog: gpu_sysman: add "LogMetrics" option

Adds "LogMetrics" option for real-time monitoring of the GPU metrics:
* Causes each metric value to be written also to collectd log, i.e. by default to console
* Avoids need to setup write plugins and delay caused by their use when investigating GPU behavior
* _Output is much more readable than write plugins' output_ [1]

[1] output looks like this (when only memory metric with just base output variant is enabled):
```
[1058954.410] b3:00.0: memory_used_bytes / device [0]: 26763264.000
[1058957.527] 3c:00.0: memory_used_bytes / device [0]: 26763264.000
[1058957.527] b3:00.0: memory_used_bytes / device [0]: 26763264.000
[1058961.528] 3c:00.0: memory_used_bytes / device [0]: 26763264.000
```

How it differs from the current write plugins' output:
* Single, human readable line per metric
* `dmesg` style "uptime-seconds" timestamp, as it's short and can be correlated with kernel output
* Sub-second timestamp accuracy for debugging which metric(s) cause "read took too long" issues
   - user can then either disable that particular metric type or increase plugin query interval
* Metric info prefixed by (abbreviated) PCI ID of the device generating that metric
  - (after logging, non-abbreviated PCI ID is added as label to metric, so write plugin could get that info too)
* Only values for specific labels are output, and not their names, to avoid line wrapping
   - Some metrics have lot of labels, and some label values can be quite long

### CODE REVIEW

**Comment by eero-t** [comment]
This is based on #4173, so it's better to wait until that's merged and I've rebased this PR on top of it, before looking at the code.

**Comment by eero-t** [comment]
@octo replying to your comment from: https://github.com/collectd/collectd/pull/4173#issuecomment-1831549443

> The [LogFile plugin](https://github.com/collectd/collectd/wiki/Plugin-LogFile) has a Timestamp option that will print the time at which the line is printed. Alternatively, the potential new "human readable" format could print the metric time in a similar fashion.

Uptime-seconds takes less space that ISO date/time, but more importantly, it's also more accurate, so option for such format would be nice.

> The label selection seems quite specific to the gpu_sysman plugin. We could add an IgnoreLabel option to write_log, but I get that configuring this every time you want to test during debugging can be annoying.

It should rather be accept than ignore list.   If it's not set, then all labels would be shown.

> Printing some labels as a prefix and others as part of the metric is also something that doesn't lend itself to a generic approach.

Being able to output most relevant _fixed size_ information first is important, as it being always in exactly the same place horizontally makes reading it _much_ easier.

Maybe options like this would work (comments list sysman plugin settings)?
```
Timestamp <false|true>       # true
TimestampFormat <iso|kernel> # kernel
LabelsBeforeMetric <list>    # "pci_bdf"
LabelsAfterMetric <list>     # "direction,location,type"
ShowLabelName <true|false>   # false
```

Label separator is output only if metric has listed label. If Labels* config values are not set, all labels are listed after metric name, none before.

Even better could be being able to set Labels* config options per metric prefix, as that way different settings could be set both for different plugins, and different metrics from those plugins:
```
<Plugin write_log>
  TimestampFormat kernel
  Timestamp true
  <MetricPrefixFormats>
     <MetricPrefix gpu_sysman_temperature>
        LabelsBeforeMetric "pci_bdf"
        LabelsAfterMetric "location"
        ShowLabelName false
     </MetricPrefix>
     <MetricPrefix gpu_sysman_membw>
        LabelsBeforeMetric "pci_bdf"
        LabelsAfterMetric "direction"
        ShowLabelName false
...
```

**Comment by octo** [comment]
* Adding a "kernel" timestamp format SGTM. As long as it's configurable – I suspect that a sizable part of the user base would find it inconvenient.
* For flexibility, both "ignore" and "accept" options may exist. Both options could be regular expressions. I would find it most intuitive if the default was "accept", i.e. to print all labels. We can evaluate "accept" first, "ignore" second, default to "accept". This way you could get your desired behavior with `Ignore "/.*/"`, `Accept "specific_label"`.
* With regards to other options I would recommend to follow the [YAGNI principle](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) and only build what you need.

**Comment by octo** [comment]
After having used the *write_log plugin* a bunch over the last few days I can confirm that it is lacking for debugging purposes. That it defaults to the graphite format is particularly annoying.

What requirements to we have for a useful format? I'd have the following ideas:

* Limit the line length so output doesn't wrap. Makes it easier for humans to consume / find what you're looking for.
* Make it possible to print fixed-width labels first.

What else?

How do you feel about multi-line output? The OpenTelemetry collector produces the following for a single metric:

```
Resource SchemaURL: 
Resource attributes:
     -> host.id: Str(911c4cdca88ae8847dbaab5b6358febf)
     -> host.name: Str(example.com)
     -> service.name: Str(unknown_service:collectd)
ScopeMetrics #0
ScopeMetrics SchemaURL: 
InstrumentationScope collectd 5.12.0.392.ga564ad0
Metric #0
Descriptor:
     -> Name: collectd_load_longterm
     -> Description: fam->help == NULL
     -> Unit: 
     -> DataType: Gauge
NumberDataPoints #0
Data point attributes:
     -> instance: Str(example.com)
StartTimestamp: 1970-01-01 00:00:00 +0000 UTC
Timestamp: 2023-12-16 23:59:22.316159357 +0000 UTC
Value: 0.861328
```

This is hard to consume, but gives excellent detail if you need to hunt down something specific. For collectd, I was thinking about an output like:

```
Resource #0: <set of resrouce attributes>
    Family #0: <metric family name>, <type>
        Metric #0: <metric labels> <timestamp> <value>
        Metric #1: <metric labels> <timestamp> <value>
        Metric #2: <metric labels> <timestamp> <value>
```

File: src/gpu_sysman.c
Comment (octo) [inline src/gpu_sysman.c:787]: I find the `m++` easy to miss. Doing

```c
metric_t *m = fam->metric.ptr + i;
```

or

```c
metric_t *m = &fam->metric.ptr[i];
```

inside the loop would be easier to grasp.

File: src/gpu_sysman.c
Comment (octo) [inline src/gpu_sysman.c:785]: It looks like the `ts` and `pci_bdf` variables are only used if `config.logmetrics` is true. I think it would be more readable if you initialized them unconditionally.

**Review by octo — APPROVED** [review]

**Comment by eero-t** [comment]
@octo Thanks for the review!  (+ sorry for accidentally editing your comment earlier, I noticed it only now...)

> After having used the _write_log plugin_ a bunch over the last few days I can confirm that it is lacking for debugging purposes. That it defaults to the graphite format is particularly annoying.

Great that you tested it!

> What requirements to we have for a useful format? I'd have the following ideas:
>     * Limit the line length so output doesn't wrap. Makes it easier for humans to consume / find what you're looking for.

No need, one can use CLI utils for that (change pipe 4K buffer to line buffer, cut output to 80 chars):
`stdbuf -oL <collectd> | cut -b-80`

Or pipe collectd output to `less -S` (to chop output to console size), and press `F` to follow the output.

> * Make it possible to print fixed-width labels first.

I don't think collectd could have reasonable logic for this, and I'd anyway would like an option listing accepted labels, so such label list could as well dictate in which order they are output.

(Only if that is not set, print all, in whatever order.)

> What else?

Whether label names are output in addition to their values.

When there's just a small subset of labels, and one knows in which order they're printed, the label names are redundant.

> How do you feel about multi-line output?

While that could be OK, I would much prefer controls for minimizing the data (labels) that is output...


> The OpenTelemetry collector produces the following for a single metric:
> ...
> This is hard to consume, but gives excellent detail if you need to hunt down something specific.

I think outputting everything is good default, as it depends a lot on the situation what parts of the data are relevant.

> For collectd, I was thinking about an output like:
> 
> ```
> Resource #0: <set of resrouce attributes>
>     Family #0: <metric family name>, <type>
>         Metric #0: <metric labels> <timestamp> <value>
>         Metric #1: <metric labels> <timestamp> <value>
>         Metric #2: <metric labels> <timestamp> <value>
> ```

If metric ordering does not change, this could work:
```
Resource #0: <set of resource attributes>
    Family #0: <metric family name>, <type>
        Metrics: <metric names>
          #0: <timestamp> <metric labels> <value>
          #1: <timestamp> <metric labels> <value>
...
```

(Metric & label names + label values differ in size, but timestamp is fixed size.)

File: src/gpu_sysman.c
Comment (eero-t) [reply src/gpu_sysman.c]: NAK.  `cdtime()` uses `REALTIME` clock (i.e. jumpy wall-clock) instead of `MONOTONIC` one.

File: src/gpu_sysman.c
Comment (eero-t) [reply src/gpu_sysman.c]: Opted for the first suggestion.

File: src/gpu_sysman.c
Comment (eero-t) [reply src/gpu_sysman.c]: Done.

File: src/gpu_sysman.c
Comment (eero-t) [reply src/gpu_sysman.c]: Great suggestion, applied.

PS. that forced me to drop the mocked metric functions from `gpu_sysman_test.c`, and include real one instead.

FYI: I've been a bit hesitant on using collectd's own functions in plugin tests instead of mocked ones, as their dependencies pull in other collected code, which fails to build with `gcc -Werror` + the extra compiler warnings [1] I'm always using for the test code.  That's why I do not use collectd build scripts for testing, and `#include` relevant collectd source files directly in the test code.

[1] For a subset of those, see e.g. https://github.com/collectd/collectd/pull/3970 and https://github.com/collectd/collectd/pull/3917.

Comment (eero-t) [reply src/gpu_sysman.c]: > that forced me to drop the mocked metric functions from gpu_sysman_test.c, and include real one instead.

While above mentioned `-Werror` tests passed with GCC in Fedora 38, they did not pass with Ubuntu 22.04 LTS => added workaround to `src/utils/metadata/meta_data.c`.

**Review by octo — APPROVED** [review]
LGTM

File: src/gpu_sysman.c
Comment (octo) [reply src/gpu_sysman.c]: I would actually recommend to prefer the real implementation over mocked or faked ones – you'll actually test the production code this way. I discovered #4214 when writing unit tests for the Write Prometheus plugin, for example.

Comment (eero-t) [reply src/gpu_sysman.c]: Do you want me to split the 3 testing related commits to a separate PR?

Comment (eero-t) [reply src/gpu_sysman.c]: I split independent testing changes to a separate PR https://github.com/collectd/collectd/pull/4223.

**Comment by octo** [comment]
Hey @eero-t, I prefer to let people merge themselves, to allow them to include small post-review fixes, such as typos or additional documentation. If you want me to merge this PR, please let me know so we don't block on each other.

**Comment by eero-t** [comment]
> Hey @eero-t, I prefer to let people merge themselves, to allow them to include small post-review fixes, such as typos or additional documentation. If you want me to merge this PR, please let me know so we don't block on each other.

Let's finish the discussion on test code PR (that I split from this) first, as it's (now) a dependency for this.

**Comment by eero-t** [comment]
No functional changes, just rebase to updated #4223 + earlier review update commit squashed (as changes were OK).  Can be merged after #4223 is in.

Conflicts with changes in #4267.  As #4267 is larger, probably better to get that in first, and update this afterwards.

**Comment by eero-t** [comment]
> Hey @eero-t, I prefer to let people merge themselves, to allow them to include small post-review fixes, such as typos or additional documentation. If you want me to merge this PR, please let me know so we don't block on each other.

@octo I'm not sure what to do with the units, once #4267 is in.  Maybe OTEL abbreviations => human readable unit names mapping from Prometheus plugin should be moved to some common code?

(I've also been keeping this open in case you'll have time at some point to add similarly readable output to `write_log` plugin, as that would obsolete this functionality.)

**Comment by eero-t** [comment]
There was a review comment in #4267 about rates and metrics logging from plugin being a "debugging" feature.  But debugging collectd metrics reporting is not the main use-case for this, just a useful side-effect (of it being done from a read plugin, so that its output content can be compared against write plugin one).

Main use-case for this feature is local GPU monitoring, an important use-case for _end-users_.

While specialized tools for that:
* https://developer.nvidia.com/nvidia-system-management-interface
* https://github.com/intel/xpumanager/blob/master/doc/CLI_user_guide.md
* https://rocm.docs.amd.com/projects/amdsmi/en/latest/amdsmi_cli_readme_link.html#usage

are naturally easier to use and have much fancier output (and much more features) than what's in this PR, I think there's still value in `collectd` being an option for such monitoring tool.

**Comment by eero-t** [comment]
@octo I have several open PRs and will open soon another.  This helps in checking that they work correctly, but overlaps with changes in them, so juggling the PRs & stacking branches is starting to be chore.

=> I'd like to merge this first.  Do the last 2 commits also have approval?

**Comment by eero-t** [comment]
Only fail is for virt plugin (on Debian unstable), i.e. unrelated.

**Comment by eero-t** [comment]
> Hey @eero-t, I prefer to let people merge themselves, to allow them to include small post-review fixes, such as typos or additional documentation. If you want me to merge this PR, please let me know so we don't block on each other.

As the changes after approval are trivial improvement, and there are no objections on them, merging this now.

(This has been going on long enough that there are several other PRs that overlap with changes here.)

**Pull request merged**
